{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ABWSE (A Better Way to Search for an Enzyme)\n",
    "\n",
    "> Yujia Liu liuyujiaokok@126.com 09/17/2018\n",
    "\n",
    "Abwse is a easier to use API that does simple job: you give the **name** of the enzyme along with the host **organism**, the API would prompt you with the most likely matching gene, EC number, sequence, pathway, chromosome coordinate, etc. for the desired enzyme.\n",
    "\n",
    "The information are accessed through NCBI.\n",
    "\n",
    "This API is only used for **pattern organisms**, which are finely sequenced and annotated.\n",
    "\n",
    "*Notice*: If there are alternative splicing of the CDS, EC number will only return the first one, but translation would give a whole list.\n",
    "\n",
    "Take *E.coli* and the enzyme **Beta-galactosidase** as an example.\n",
    "\n",
    "Also, it is normal that inquiry of the EC number of eukaryotic enzymes, for example, of human, is always empty."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+------------------+-----------------+----------------------+------------------+--------------------+\n",
      "|       name       |   enzyme_name   |       species        |      start       |        end         |\n",
      "+------------------+-----------------+----------------------+------------------+--------------------+\n",
      "|       SIGE       | sigma factor E  | Arabidopsis thaliana |     8157151      |      8160343       |\n",
      "|                  |                 |     (thale cress)    |                  |                    |\n",
      "+------------------+-----------------+----------------------+------------------+--------------------+\n",
      "|   description    |   chromosome    |          EC          |      nt_seq      |      tran_seq      |\n",
      "+------------------+-----------------+----------------------+------------------+--------------------+\n",
      "| Encodes a specia |        5        |         None         | TATAGGTTTTCTTCAG | 3 types; MGVVSISSS |\n",
      "| lized sigma fa.. |                 |                      | TATTCACCATGACT.. | AARSPLGLGNDLLTHR.. |\n",
      "|        .         |                 |                      |        .         |         .          |\n",
      "+------------------+-----------------+----------------------+------------------+--------------------+\n",
      "|       name       |   enzyme_name   |       species        |      start       |        end         |\n",
      "+------------------+-----------------+----------------------+------------------+--------------------+\n",
      "|       lacZ       | beta-D-galactos | Escherichia coli str |      363231      |       366305       |\n",
      "|                  |      idase      | . K-12 substr. MG165 |                  |                    |\n",
      "|                  |                 |          5           |                  |                    |\n",
      "+------------------+-----------------+----------------------+------------------+--------------------+\n",
      "|   description    |   chromosome    |          EC          |      nt_seq      |      tran_seq      |\n",
      "+------------------+-----------------+----------------------+------------------+--------------------+\n",
      "| enzyme; Degradat |      None       |       3.2.1.23       | TTATTTTTGACACCAG | 1 types; MTMITDSLA |\n",
      "| ion of small m.. |                 |                      | ACCAACTGGTAATG.. | VVLQRRDWENPGVTQL.. |\n",
      "|        .         |                 |                      |        .         |         .          |\n",
      "+------------------+-----------------+----------------------+------------------+--------------------+\n",
      "|       name       |   enzyme_name   |       species        |      start       |        end         |\n",
      "+------------------+-----------------+----------------------+------------------+--------------------+\n",
      "|      ACAT2       |  acetyl-CoA ace | Homo sapiens (human) |    159761957     |     159779112      |\n",
      "|                  | tyltransferase  |                      |                  |                    |\n",
      "|                  |        2        |                      |                  |                    |\n",
      "+------------------+-----------------+----------------------+------------------+--------------------+\n",
      "|   description    |   chromosome    |          EC          |      nt_seq      |      tran_seq      |\n",
      "+------------------+-----------------+----------------------+------------------+--------------------+\n",
      "| superoxide dismu |        6        |         None         | GGTGCGCGGGGAGGTG | 4 types; MNAGSDPVV |\n",
      "| tase 2; Derive.. |                 |                      | GAGGGCGAGGGGCG.. | IVSAARTIIGSFNGAL.. |\n",
      "|        .         |                 |                      |        .         |         .          |\n",
      "+------------------+-----------------+----------------------+------------------+--------------------+\n",
      "|       name       |   enzyme_name   |       species        |      start       |        end         |\n",
      "+------------------+-----------------+----------------------+------------------+--------------------+\n",
      "|        Cs        |  citrate syntha | Mus musculus (house  |    128337832     |     128362479      |\n",
      "|                  |       se        |        mouse)        |                  |                    |\n",
      "+------------------+-----------------+----------------------+------------------+--------------------+\n",
      "|   description    |   chromosome    |          EC          |      nt_seq      |      tran_seq      |\n",
      "+------------------+-----------------+----------------------+------------------+--------------------+\n",
      "| citrate synthase |       10        |         None         | CTCCATCACAGCGGCG | 1 types; MALLTAATR |\n",
      "| ; Derived by a.. |                 |                      | ACCGCTCCTGTTGC.. | LLGAKNSSCLVLAARH.. |\n",
      "|        .         |                 |                      |        .         |         .          |\n",
      "+------------------+-----------------+----------------------+------------------+--------------------+\n"
     ]
    }
   ],
   "source": [
    "import abwse\n",
    "\n",
    "enzyme_list = []\n",
    "\n",
    "enzyme_info = abwse.ezsearch('DNA polymerase sigma', 'Arabidopsis thaliana', 'rainl199922@berkeley.edu', rettype='dict')\n",
    "enzyme_list.append(enzyme_info)\n",
    "\n",
    "enzyme_info = abwse.ezsearch('Beta-galactosidase', 'Escherichia coli', 'rainl199922@berkeley.edu', rettype='dict')\n",
    "enzyme_list.append(enzyme_info)\n",
    "\n",
    "enzyme_info = abwse.ezsearch('Acetyl CoA transferase', 'Homo Sapiens', 'rainl199922@berkeley.edu', rettype='dict')\n",
    "enzyme_list.append(enzyme_info)\n",
    "\n",
    "enzyme_info = abwse.ezsearch('Citrate synthase', 'Mus musculus', 'rainl199922@berkeley.edu', rettype='dict')\n",
    "enzyme_list.append(enzyme_info)\n",
    "\n",
    "print(abwse.eztable(*enzyme_list, width=100))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Core coding is listed here:\n",
    "\n",
    "```py\n",
    "from Bio import Entrez, SeqIO\n",
    "from beautifultable import BeautifulTable as BT\n",
    "import re, sys, json\n",
    "\n",
    "# function to index details in genbank features. note this function return a list\n",
    "def __index_genbank_features__(gb_record, feature_type, qualifier) :\n",
    "    answer = []\n",
    "    for (index, feature) in enumerate(gb_record.features) :\n",
    "        if feature.type==feature_type :\n",
    "            if qualifier in feature.qualifiers :\n",
    "                for value in feature.qualifiers[qualifier] :\n",
    "                    answer.append(value)\n",
    "    return answer\n",
    "\n",
    "def ezsearch(enzyme, organism, email, rettype='json', pathprefix='./ezsearch_ret'):\n",
    "\n",
    "    # provide email for entrez\n",
    "    Entrez.email = email\n",
    "\n",
    "    # look up for the gene name and annotation from NCBI\n",
    "    term_str = \"{}[orgn] AND {}\".format(organism, enzyme)\n",
    "    search_ret = Entrez.esearch(db='gene', term=term_str, sort='relevance')\n",
    "    try:\n",
    "        id_first = Entrez.read(search_ret)['IdList'][0]\n",
    "    except IndexError:\n",
    "        print(\"No match for that!\")\n",
    "        return None\n",
    "    gene_info = Entrez.efetch(db='gene', id=id_first, retmode='text').read()\n",
    "    # following is a piece of example of return for a gene query\n",
    "    \"\"\"\n",
    "\n",
    "    1. lacZ\n",
    "    beta-D-galactosidase [Escherichia coli str. K-12 substr. MG1655]\n",
    "    Other Aliases: b0344, ECK0341, JW0335\n",
    "    Annotation:  NC_000913.3 (363231..366305, complement)\n",
    "    ID: 945006\n",
    "    \"\"\"\n",
    "    gene_info_lines = gene_info.split('\\n')\n",
    "    gene_name = gene_info_lines[1][3:]\n",
    "    try:\n",
    "        gene_enzyme_name = re.findall(r\"[ a-zA-Z0-9-]+(?= \\[)\", gene_info_lines[2])[0]\n",
    "        gene_species = re.findall(r\"(?<=\\[)[ a-zA-Z0-9-\\.\\(\\):]+(?=\\])\", gene_info_lines[2])[0]\n",
    "        gene_annotation = re.findall(r\"(?<=Annotation: )[ A-Za-z0-9\\._]+\", gene_info)[0]\n",
    "        gene_start = re.findall(r\"(?<=\\()[\\d]+(?=\\.\\.)\", gene_info)[0]\n",
    "        gene_end = re.findall(r\"(?<=\\.\\.)[\\d]+\", gene_info)[0]\n",
    "    except IndexError:\n",
    "        print(\"Error in parsing the gene informatin. Here is the info from NCBI:{}\".format(gene_info))\n",
    "        return None\n",
    "\n",
    "    # look up for whole sequence and other stuff according to the chromosome coordinate\n",
    "    search_ret = Entrez.esearch(db='nucleotide', term=gene_annotation, sort='relevance')\n",
    "    try:\n",
    "        id_first = Entrez.read(search_ret)['IdList'][0]\n",
    "    except IndexError:\n",
    "        print(\"Matches for gene, but no match for the genome position!\")\n",
    "        return None\n",
    "    gb_ret = Entrez.efetch(db='nuccore', id=id_first,\\\n",
    "                           seq_start=gene_start, seq_stop=gene_end, rettype='gb', retmode='text').read()\n",
    "\n",
    "    # store genbank file\n",
    "    gb_file = open(pathprefix+'.gb', 'w')\n",
    "    gb_file.writelines(gb_ret)\n",
    "    gb_file.close()\n",
    "\n",
    "    # get other detailed info from genbank file\n",
    "    gb_parse = SeqIO.parse(pathprefix+'.gb', 'gb')\n",
    "    gb_record = list(gb_parse)[0]\n",
    "    gene_chromosome = __index_genbank_features__(gb_record, 'source', 'chromosome')\n",
    "    if len(gene_chromosome) == 0:\n",
    "        gene_chromosome = None\n",
    "    else:\n",
    "        gene_chromosome = gene_chromosome[0]    #only one chromosome possible\n",
    "    gene_note = __index_genbank_features__(gb_record, 'gene', 'note')\n",
    "    if len(gene_note) == 0:\n",
    "        gene_note = None\n",
    "    else:\n",
    "        gene_note = gene_note[0]\n",
    "    gene_description = __index_genbank_features__(gb_record, 'CDS', 'function')\n",
    "    if len(gene_description) == 0:\n",
    "        gene_description = gene_note\n",
    "    else:\n",
    "        gene_description = gene_description[0]\n",
    "    gene_EC = __index_genbank_features__(gb_record, 'CDS', 'EC_number')\n",
    "    if len(gene_EC) == 0:\n",
    "        gene_EC = None\n",
    "    else:\n",
    "        gene_EC = gene_EC[0]    #only one chromosome possible\n",
    "    gene_nt_seq = gb_record.seq\n",
    "    gene_tran_seq = __index_genbank_features__(gb_record, 'CDS', 'translation')\n",
    "    #due to alternative splicing, there could be multiple translations\n",
    "\n",
    "    # store all query data to dict\n",
    "    gene = {}\n",
    "    gene['name'] = gene_name\n",
    "    gene['enzyme_name'] = gene_enzyme_name\n",
    "    gene['species'] = gene_species\n",
    "    gene['start'] = gene_start\n",
    "    gene['end'] = gene_end\n",
    "    gene['description'] = gene_description\n",
    "    gene['chromosome'] = gene_chromosome\n",
    "    gene['EC'] = gene_EC\n",
    "    gene['nt_seq'] = str(gene_nt_seq)\n",
    "    gene['tran_seq'] = gene_tran_seq\n",
    "\n",
    "    # return required format\n",
    "    if rettype == 'json':\n",
    "        return json.dumps(gene)\n",
    "    elif rettype == 'dict':\n",
    "        return gene\n",
    "    else:\n",
    "        print('Wrong returning format! Returning json instead!')\n",
    "        return json.dumps(gene)\n",
    "\n",
    "def eztable(enzyme_info, width=100) :\n",
    "    enzyme_table = BT(max_width=width)\n",
    "    enzyme_table.append_row(['name', 'enzyme_name', 'species', 'start', 'end'])\n",
    "    enzyme_table.append_row([enzyme_info['name'],\\\n",
    "                            enzyme_info['enzyme_name'],\\\n",
    "                            enzyme_info['species'],\\\n",
    "                            enzyme_info['start'],\\\n",
    "                            enzyme_info['end']])\n",
    "    enzyme_table.append_row(['description', 'chromosome', 'EC', 'nt_seq', 'tran_seq'])\n",
    "    try:\n",
    "        description = enzyme_info['description']\n",
    "        enzyme_table.append_row([description if description is None or len(description)<30 else description[:30]+'...',\\\n",
    "                               enzyme_info['chromosome'],\\\n",
    "                               enzyme_info['EC'],\\\n",
    "                               enzyme_info['nt_seq'][:30]+'...',\\\n",
    "                               str(len(enzyme_info['tran_seq']))+' types; '+enzyme_info['tran_seq'][0][:25]+'...'])\n",
    "    except:\n",
    "        print('Unexpected data structure! May lack in import field!')\n",
    "        return None\n",
    "    return enzyme_table\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
